opentelemetry-sdk/tests/events/test_events.py

```diff
@@ -0,0 +1,162 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=protected-access,no-self-use
+
+import unittest
+from unittest.mock import Mock, patch
+
+from opentelemetry._events import Event
+from opentelemetry._logs import SeverityNumber, set_logger_provider
+from opentelemetry.sdk._events import EventLoggerProvider
+from opentelemetry.sdk._logs import LoggerProvider
+from opentelemetry.sdk._logs._internal import Logger, NoOpLogger
+from opentelemetry.sdk.environment_variables import OTEL_SDK_DISABLED
+
+
+class TestEventLoggerProvider(unittest.TestCase):
```

Comment:
Nice to see thorough testing here, and this is not a blocking comment, but would you characterize these as testing implementation or testing behavior? In other words, if the internals of `EventLoggerProvider` were refactored, but the class worked as it should, would these tests pass? Conversely, if the external behavior of the class changed but the code retained the internal implementation as tested, would these tests fail?

Also, perhaps this isn't the primary goal of testing, but how instructive are these tests to other developers as to how to use these classes? Is it worthwhile to attempt to use real types and the API?

Reply:
@pmcollins 

These are good questions. I think we should have a mixture of both tests depending on the specific api/sdk we are testing. For components that have specific implementation detail outlined by the spec (e.g. [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-sdk.md#eventloggerprovider) `The EventLoggerProvider MUST be implemented as a proxy to an instance of [LoggerProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerprovider).`), these would be an example of having a test that tests implementation detail. However in all other cases I believe we should be testing behavior since this is more useful for catching regressions. I'm not sure if all of our tests across our components are consistent in this philosophy currently, we did not have any strict guidance for this area.

> how instructive are these tests to other developers as to how to use these classes?

I would argue currently it is not as intuitive for other developers the purpose of our testing. We might need a complete audit of our testing infrastructure/test suite as a whole. Might be a bigger discussion however.